File: packages/core/src/engine/oracle.ts
import { MULT_Q6 } from '../constants';
import { MathError, refPriceQ8, toQ8 } from '../math/fixed';
import { ErrorCode, type ErrorCode as ErrorCodeT } from './errors';
import type { ArenaAssetSpec, PriceMode, SidePriceSnapshot } from './types';

/**
 * Oracle input validation (ARCHITECTURE §5.3, ARENA_STATE_MACHINE §5.1–5.2).
 * Mirrors the checks the program performs on a Pyth `PriceUpdateV2` account.
 */

export type VerificationLevel = 'Full' | 'Partial';

export interface PriceInput {
  feedId: string;
  price: bigint;
  /** Pyth confidence interval, same exponent as price. */
  conf: bigint;
  expo: number;
  publishTime: bigint;
  verificationLevel: VerificationLevel;
  /** For ScaledUi mints: multiplier read from the mint at snapshot time (Q6). */
  multQ6?: bigint;
}

export type OracleCheck =
  { ok: true; snapshot: SidePriceSnapshot } | { ok: false; code: ErrorCodeT; detail: string };

/**
 * Validate a price update against the asset policy for a target timestamp.
 *
 * Exact mode:     |publish − target| ≤ tolerance
 * LastKnown mode: target − maxClosedStaleness ≤ publish ≤ target  (equities
 *                 outside market hours; requires allowClosed)
 */
export function validatePriceUpdate(
  input: PriceInput,
  asset: ArenaAssetSpec,
  targetTs: bigint,
  allowClosed: boolean,
): OracleCheck {
  if (input.feedId.toLowerCase() !== asset.feedId.toLowerCase()) {
    return {
      ok: false,
      code: ErrorCode.OracleFeedMismatch,
      detail: `${input.feedId} ≠ ${asset.feedId}`,
    };
  }
  if (input.verificationLevel !== 'Full') {
    return { ok: false, code: ErrorCode.OracleNotFullyVerified, detail: input.verificationLevel };
  }
  if (input.price <= 0n) {
    return { ok: false, code: ErrorCode.OracleNonPositivePrice, detail: `${input.price}` };
  }
  if (input.conf < 0n) {
    return { ok: false, code: ErrorCode.OracleConfidenceTooWide, detail: 'negative conf' };
  }
  // conf × 10_000 / price ≤ maxConfBps  (same exponent, so no normalisation needed)
  if ((input.conf * 10_000n) / input.price > asset.maxConfBps) {
    return {
      ok: false,
      code: ErrorCode.OracleConfidenceTooWide,
      detail: `conf ${input.conf} / price ${input.price}`,
    };
  }

  let mode: PriceMode;
  const diff = input.publishTime - targetTs;
  const absDiff = diff < 0n ? -diff : diff;
  if (absDiff <= asset.toleranceSecs) {
    mode = 'Exact';
  } else if (diff > 0n) {
    return { ok: false, code: ErrorCode.OracleTooEarly, detail: `published ${diff}s after target` };
  } else {
    const lastKnownAllowed =
      allowClosed && asset.assetClass !== 'Crypto' && asset.maxClosedStalenessSecs > 0n;
    if (!lastKnownAllowed || absDiff > asset.maxClosedStalenessSecs) {
      return {
        ok: false,
        code: ErrorCode.OracleStale,
        detail: `published ${absDiff}s before target`,
      };
    }
    mode = 'LastKnown';
  }

  let oraclePriceQ8: bigint;
  try {
    oraclePriceQ8 = toQ8(input.price, input.expo);
  } catch (e) {
    return {
      ok: false,
      code: e instanceof MathError ? ErrorCode.OracleUnsupportedExponent : ErrorCode.MathOverflow,
      detail: e instanceof Error ? e.message : String(e),
    };
  }

  const multQ6 = asset.scaledUi ? (input.multQ6 ?? MULT_Q6) : MULT_Q6;
  if (multQ6 <= 0n) {
    return { ok: false, code: ErrorCode.InvalidParams, detail: 'multiplier must be > 0' };
  }
  let priceQ8: bigint;
  try {
    priceQ8 = refPriceQ8(oraclePriceQ8, multQ6);
  } catch (e) {
    return {
      ok: false,
      code: ErrorCode.MathOverflow,
      detail: e instanceof Error ? e.message : String(e),
    };
  }

  return {
    ok: true,
    snapshot: { priceQ8, oraclePriceQ8, publishTime: input.publishTime, mode, multQ6 },
  };
}

File: programs/tribe_arena/src/engine/oracle.rs
//! Oracle input validation — transcription of `packages/core/src/engine/oracle.ts`.
//! Works on already-parsed fields so it is testable without a live Pyth
//! account; the instruction layer extracts them from `PriceUpdateV2`.

use crate::engine::math::{ref_price_q8, to_q8, MULT_Q6};
use crate::errors::TribeError;
use crate::state::{asset_class, price_mode, ArenaAsset, PriceSnapshot};
use anchor_lang::prelude::*;

#[derive(Clone, Copy, Debug)]
pub struct PriceInput {
    pub feed_id: [u8; 32],
    pub price: i64,
    pub conf: u64,
    pub expo: i32,
    pub publish_time: i64,
    pub fully_verified: bool,
    /// Scaled-UI multiplier read from the mint (Q6); ignored for non-scaled assets.
    pub mult_q6: Option<u64>,
}

/// Exact mode: |publish − target| ≤ tolerance.
/// LastKnown mode (non-crypto, allowed): target − max_closed ≤ publish ≤ target.
pub fn validate_price_update(
    input: &PriceInput,
    asset: &ArenaAsset,
    target_ts: i64,
    allow_closed: bool,
) -> Result<PriceSnapshot> {
    require!(input.feed_id == asset.feed_id, TribeError::OracleFeedMismatch);
    require!(input.fully_verified, TribeError::OracleNotFullyVerified);
    require!(input.price > 0, TribeError::OracleNonPositivePrice);
    // conf × 10_000 / price ≤ max_conf_bps (same exponent)
    let conf_bps = (input.conf as u128) * 10_000 / (input.price as u128);
    require!(
        conf_bps <= asset.max_conf_bps as u128,
        TribeError::OracleConfidenceTooWide
    );

    let diff = input.publish_time - target_ts;
    let abs_diff = diff.abs();
    let mode = if abs_diff <= asset.tolerance_secs {
        price_mode::EXACT
    } else if diff > 0 {
        return err!(TribeError::OracleTooEarly);
    } else {
        let last_known_allowed = allow_closed
            && asset.asset_class != asset_class::CRYPTO
            && asset.max_closed_staleness_secs > 0;
        require!(
            last_known_allowed && abs_diff <= asset.max_closed_staleness_secs,
            TribeError::OracleStale
        );
        price_mode::LAST_KNOWN
    };

    let oracle_price_q8 = to_q8(input.price, input.expo)?;
    let mult_q6 = if asset.scaled_ui {
        input.mult_q6.unwrap_or(MULT_Q6 as u64)
    } else {
        MULT_Q6 as u64
    };
    require!(mult_q6 > 0, TribeError::InvalidParams);
    let price_q8 = ref_price_q8(oracle_price_q8, mult_q6)?;
    Ok(PriceSnapshot {
        price_q8,
        oracle_price_q8,
        publish_time: input.publish_time,
        mode,
        mult_q6,
    })
}
